f3/granite.go

```diff
@@ -985,10 +984,65 @@ func (q *quorumState) ListStrongQuorumAgreedValues() []ECChain {
 			withQuorum = append(withQuorum, q.chainSupport[cid].chain)
 		}
 	}
-	sortByWeight(withQuorum)
 	return withQuorum
 }
 
+//// QUALITY phase helper /////
+
+type qualityQuorumState struct {
+	quorumState
+	// Map to store the smallest ticket for each chain value.
+	smallestTickets map[TipSetID]Ticket
+}
+
+func newQualityQuorumState(powerTable PowerTable) *qualityQuorumState {
+	return &qualityQuorumState{
+		quorumState:     *newQuorumState(powerTable),
+		smallestTickets: make(map[TipSetID]Ticket),
+	}
+}
+
+// Receive overrides the Receive method of quorumState to handle the smallest ticket per value logic.
+func (q *qualityQuorumState) Receive(sender ActorID, value ECChain, signature []byte, justification Justification, ticket Ticket) {
+	head := value.HeadCIDOrZero()
+
+	q.quorumState.Receive(sender, value, signature, justification)
+
+	// Update the smallest ticket
+	currentTicket, exists := q.smallestTickets[head]
+	if !exists || (ticket != nil && currentTicket.Compare(ticket) > 0) {
+		q.smallestTickets[head] = ticket
+	}
+}
+
+// ListStrongQuorumAgreedValues lists all values sorted by the ascending
+// order of the smallest VRF ticket associated to the value
+func (q *qualityQuorumState) ListStrongQuorumAgreedValues() []ECChain {
+	var withQuorum []ECChain
+	for cid, cp := range q.chainSupport {
+		if cp.hasStrongQuorum {
+			withQuorum = append(withQuorum, q.chainSupport[cid].chain)
+		}
+	}
+	sortByTicket(withQuorum, q.smallestTickets)
```

Comment:
Ah rats! We overlooked this in discussion. The QUALITY phase really does need weight because multiple strong quora are possible (over prefixes of the proposals) and we really do want to select the highest weight one (longest, for sure, and then heaviest among equal lengths).

I think ignoring weight and aborting on multiple strong quora is still the right approach for all the other phases but I think we might need tipset weight in here for the quality phase.

That is really annoying because #10 (fixing which will add some complexity). We could order on (length, CID) instead, but it would cause finalisation of non-heaviest tipset sometimes, which would be a bit nasty. (Using a ticket would be hardly better).

Reply:
Done, and actually did much more than that after I saw how we are selecting the prefix in QUALITY. See #64 for more info. Pushed.